translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,405 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opencensus
+
+import (
+	"strings"
+
+	occommon "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	ocresource "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	octrace "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/timestamp"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+// OTLP attributes to map certain OpenCensus proto fields. These fields don't have
+// corresponding fields in OTLP, nor are defined in OTLP semantic conventions.
+// TODO: decide if any of these must be in OTLP semantic conventions.
+const (
+	ocAttributeProcessStartTime  = "opencensus.starttime"
+	ocAttributeProcessID         = "opencensus.pid"
+	ocAttributeExporterVersion   = "opencensus.exporterversion"
+	ocAttributeResourceType      = "opencensus.resourcetype"
+	ocTimeEventMessageEventType  = "opencensus.timeevent.messageevent.type"
+	ocTimeEventMessageEventID    = "opencensus.timeevent.messageevent.id"
+	ocTimeEventMessageEventUSize = "opencensus.timeevent.messageevent.usize"
+	ocTimeEventMessageEventCSize = "opencensus.timeevent.messageevent.csize"
+)
+
+func ocToInternal(td consumerdata.TraceData) data.TraceData {
+
+	if td.Node == nil && td.Resource == nil && len(td.Spans) == 0 {
+		return data.TraceData{}
+	}
+
+	resource := ocNodeResourceToInternal(td.Node, td.Resource)
+
+	resourceSpans := &data.ResourceSpans{}
+	resourceSpans.SetResource(resource)
+	resourceSpanList := []*data.ResourceSpans{resourceSpans}
+
+	spanCount := len(td.Spans)
+
+	if spanCount != 0 {
+		// Create slice that holds pointers and another slice that holds structs at once
+		// to avoid individual allocations of structs.
+		content := data.NewSpanSlice(spanCount)
+		ptrs := make([]*data.Span, 0, spanCount)
+
+		for i, ocSpan := range td.Spans {
+			if ocSpan == nil {
+				// Skip nil spans.
+				continue
+			}
+
+			// Point one element in slice of pointers to an element in slice of structs.
+			dest := &content[i]
+
+			ocSpanToInternal(dest, ocSpan)
+
+			if ocSpan.Resource != nil {
+				// Add a separate ResourceSpans item just for this span since it
+				// has a different Resource.
+				separateRS := &data.ResourceSpans{}
```

Comment:
Any reason not to use https://github.com/open-telemetry/opentelemetry-collector/blob/c0f4a5a70d825c2466f98034bcfed9293a61c870/internal/data/trace.go#L56 here?
e.g.
```
separateRS := data.NewResourceSpans(
  ocNodeResourceToInternal(nil, ocSpan.Resource),
  []*data.Span{dest})
resourceSpanList = append(resourceSpanList, separateRS)
```
or you may even call it inline without extra local variable.

Reply:
Good point.

Reply:
Done.